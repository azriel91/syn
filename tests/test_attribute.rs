extern crate syn;

mod features;

#[macro_use]
mod macros;

use syn::parse::Parser;
use syn::Attribute;

#[test]
fn test_meta_item_path() {
    test("#[foo]")
}

#[test]
fn test_meta_item_path_multiple_segments() {
    test("#[foo::bar]")
}

#[test]
fn test_meta_item_path_leading_colon() {
    test("#[::foo::bar]")
}

#[test]
fn test_meta_item_name_value() {
    test("#[foo = 5]")
}

#[test]
fn test_meta_item_bool_value() {
    test("#[foo = true]");
    test("#[foo = false]")
}

#[test]
fn test_meta_item_name_value_multiple_segments() {
    test("#[foo::bar = 5]")
}

#[test]
fn test_meta_item_name_value_leading_colon() {
    test("#[::foo::bar = 5]")
}

#[test]
fn test_meta_item_list_lit() {
    test("#[foo(5)]")
}

#[test]
fn test_meta_item_list_path() {
    test("#[foo(bar::Baz)]")
}

#[test]
fn test_meta_item_list_path_multiple_segments() {
    test("#[foo::bar(baz::Qux)]")
}

#[test]
fn test_meta_item_list_path_leading_colon() {
    test("#[::foo::bar(::baz::Qux)]")
}

#[test]
fn test_meta_item_list_name_value() {
    test("#[foo(bar = 5)]")
}

#[test]
fn test_meta_item_list_name_value_multiple_segments() {
    test("#[foo(bar::baz = 5)]")
}

#[test]
fn test_meta_item_list_bool_value() {
    test("#[foo(bar = true)]")
}

#[test]
fn test_meta_item_list_bool_value_multiple_segments() {
    test("#[foo(bar::baz = true)]")
}

#[test]
fn test_meta_item_multiple() {
    test("#[foo(word, name = 5, list(name2 = 6), word2)]")
}

#[test]
fn test_meta_item_multiple_with_path() {
    test("#[foo::bar(word, name = 5, list(name2 = 6), word2, ::path::true)]")
}

#[test]
fn test_bool_lit() {
    test("#[foo(true)]")
}

fn test(input: &str) {
    let attrs = Attribute::parse_outer.parse_str(input).unwrap();

    assert_eq!(attrs.len(), 1);

    let attr = attrs.into_iter().next().unwrap();
    let interpret = snapshot!(attr.interpret_meta().unwrap());
    let parse = snapshot!(attr.parse_meta().unwrap());

    assert_eq!(interpret, parse);
}
