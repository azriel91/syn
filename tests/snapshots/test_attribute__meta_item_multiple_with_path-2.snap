---
created: "2019-03-17T21:23:01.656303800Z"
creator: insta@0.7.1
source: tests/test_attribute.rs
expression: syntax_tree
---
List(
    MetaList {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: Ident(
                        foo
                    ),
                    arguments: None
                },
                Colon2,
                PathSegment {
                    ident: Ident(
                        bar
                    ),
                    arguments: None
                }
            ]
        },
        paren_token: Paren,
        nested: [
            Meta(
                Path(
                    Path {
                        leading_colon: None,
                        segments: [
                            PathSegment {
                                ident: Ident(
                                    word
                                ),
                                arguments: None
                            }
                        ]
                    }
                )
            ),
            Comma,
            Meta(
                NameValue(
                    MetaNameValue {
                        path: Path {
                            leading_colon: None,
                            segments: [
                                PathSegment {
                                    ident: Ident(
                                        name
                                    ),
                                    arguments: None
                                }
                            ]
                        },
                        eq_token: Eq,
                        lit: Int(
                            LitInt {
                                token: Literal {
                                    lit: 5
                                }
                            }
                        )
                    }
                )
            ),
            Comma,
            Meta(
                List(
                    MetaList {
                        path: Path {
                            leading_colon: None,
                            segments: [
                                PathSegment {
                                    ident: Ident(
                                        list
                                    ),
                                    arguments: None
                                }
                            ]
                        },
                        paren_token: Paren,
                        nested: [
                            Meta(
                                NameValue(
                                    MetaNameValue {
                                        path: Path {
                                            leading_colon: None,
                                            segments: [
                                                PathSegment {
                                                    ident: Ident(
                                                        name2
                                                    ),
                                                    arguments: None
                                                }
                                            ]
                                        },
                                        eq_token: Eq,
                                        lit: Int(
                                            LitInt {
                                                token: Literal {
                                                    lit: 6
                                                }
                                            }
                                        )
                                    }
                                )
                            )
                        ]
                    }
                )
            ),
            Comma,
            Meta(
                Path(
                    Path {
                        leading_colon: None,
                        segments: [
                            PathSegment {
                                ident: Ident(
                                    word2
                                ),
                                arguments: None
                            }
                        ]
                    }
                )
            ),
            Comma,
            Meta(
                Path(
                    Path {
                        leading_colon: Some(
                            Colon2
                        ),
                        segments: [
                            PathSegment {
                                ident: Ident(
                                    path
                                ),
                                arguments: None
                            },
                            Colon2,
                            PathSegment {
                                ident: Ident(
                                    true
                                ),
                                arguments: None
                            }
                        ]
                    }
                )
            )
        ]
    }
)
